executor/server.go

```diff
@@ -106,33 +107,37 @@ func (s *Server) Start(ctx context.Context) error {
 		Capability: 100,
 	}
 
-	resp, err := s.cli.RegisterExecutor(ctx, registerReq)
+	resp, err := s.cli.RegisterExecutor(ctx, registerReq, s.cfg.RPCTimeout)
 	if err != nil {
 		return err
 	}
 	s.ID = model.ExecutorID(resp.ExecutorId)
 	log.L().Logger.Info("register successful", zap.Int32("id", int32(s.ID)))
 
 	// Start Heartbeat
-	ticker := time.NewTicker(time.Duration(s.cfg.KeepAliveInterval) * time.Millisecond)
+	ticker := time.NewTicker(s.cfg.KeepAliveInterval)
+	s.lastHearbeatTime = time.Now()
 	for {
 		select {
 		case <-ctx.Done():
 			return nil
 		case <-exitCh:
 			return nil
 		case t := <-ticker.C:
+			if s.lastHearbeatTime.Add(s.cfg.KeepAliveTTL).Before(time.Now()) {
+				return errors.New("heartbeat timeout")
+			}
 			req := &pb.HeartbeatRequest{
 				ExecutorId: int32(s.ID),
 				Status:     int32(model.Running),
 				Timestamp:  uint64(t.Unix()),
-				Ttl:        uint64(s.cfg.KeepAliveTTL),
+				// We set longer ttl for master.
+				Ttl:        uint64(s.cfg.KeepAliveTTL.Milliseconds() + s.cfg.RPCTimeout.Milliseconds()),
 			}
-			// FIXME: set timeout.
-			resp, err := s.cli.SendHeartbeat(ctx, req)
+			resp, err := s.cli.SendHeartbeat(ctx, req, s.cfg.RPCTimeout)
 			if err != nil {
 				log.L().Error("heartbeat meet error", zap.Error(err))
-				if s.lastHearbeatTime.Add(time.Duration(s.cfg.KeepAliveTTL) * time.Millisecond).Before(time.Now()) {
+				if s.lastHearbeatTime.Add(s.cfg.KeepAliveTTL).Before(time.Now()) {
 					return err
```

Comment:
Should we return `heartbeat timeout` or wrapping `err` with `heartbeat time`

Reply:
Yeah, that's a good point!